storage/kvstore.go

```diff
@@ -0,0 +1,284 @@
+package storage
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"log"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/storage/backend"
+	"github.com/coreos/etcd/storage/storagepb"
+)
+
+var (
+	batchLimit    = 10000
+	batchInterval = 100 * time.Millisecond
+	keyBucketName = []byte("key")
+
+	ErrTnxIDMismatch = errors.New("storage: tnx id mismatch")
+)
+
+type store struct {
+	mu sync.RWMutex
+
+	b       backend.Backend
+	kvindex index
+
+	currentIndex uint64
+	subIndex     uint32 // tracks next subIndex to put into backend
+
+	tmu   sync.Mutex // protect the tnxID field
+	tnxID int64      // tracks the current tnxID to verify tnx operations
+}
+
+func newStore(path string) KV {
+	s := &store{
+		b:            backend.New(path, batchInterval, batchLimit),
+		kvindex:      newTreeIndex(),
+		currentIndex: 0,
+	}
+
+	tx := s.b.BatchTx()
+	tx.Lock()
+	tx.UnsafeCreateBucket(keyBucketName)
+	tx.Unlock()
+	s.b.ForceCommit()
+
+	return s
+}
+
+func (s *store) Put(key, value []byte) int64 {
+	id := s.TnxBegin()
+	s.put(key, value, s.currentIndex+1)
+	s.TnxEnd(id)
+
+	return int64(s.currentIndex)
+}
+
+func (s *store) Range(key, end []byte, limit, rangeIndex int64) (kvs []storagepb.KeyValue, index int64) {
+	id := s.TnxBegin()
+	kvs, index = s.rangeKeys(key, end, limit, rangeIndex)
+	s.TnxEnd(id)
```

Comment:
maybe we could not use tnx for Range in the future? It looks like that we do some index logic only for this case, and we will support `Range` and `Put/Delete` happen at the same time.